lib/rules/camelcase.js

```diff
@@ -125,6 +125,38 @@ module.exports = {
             return false;
         }
 
+        /**
+         * Checks whether the given node represents assignment target property in destructuring.
+         *
+         * For examples:
+         *    ({a: b.foo} = c);  // => true for `foo`
+         *    ({a: b[foo]} = c); // => true for `foo`
+         *    ([a.foo] = b);     // => true for `foo`
+         *    ([a[foo]] = b);    // => true for `foo`
+         *    ({...a.foo} = b);  // => true for `foo`
+         *    ({...a[foo]} = b); // => true for `foo`
+         * @param {ASTNode} node An Identifier node to check
+         * @returns {boolean} True if the node is an assignment target property in destructuring.
```

Comment:
I think it should also return `true` for cases such as `[a.foo = 1] = b`, otherwise this would be a false negative:

```js
/* eslint camelcase: ["error", { properties: "always" }] */

[a.b_c = 1] = d;
```